<?xml version="1.0" encoding="UTF-8"?>
<launch>
    <arg name="use_rviz" default="True" />
    <arg name="use_camera" default="True" />
    <arg name="use_bag" default="False" />
    <arg name="image_width" default="1920" /> 
    <arg name="image_height" default="1080" />
<!--     <arg name="image_width" default="640" /> 
    <arg name="image_height" default="480" /> -->
    <arg name="camera" default="/usb_cam/image_raw" />
    <arg name="camera_info" default="/camera/camera_info" />
    <arg name="tf_prefix" default="marker_id" />
    <arg name="show_detections" default="true" />
    <arg name="marker_size" default="2" doc="length of the physical marker's side in meters"/>
    <!-- <arg name="dictionary_name" default="DICT_5X5_50" /> -->
    <arg name="dictionary_name" default="DICT_6X6_250" />
    <arg name="blur_window_size" default="7" />
    <arg name="num_detected" default="100" /><!-- 0: do not use it, -->
    <arg name="min_prec_value" default="50" /><!-- 0: do not use it, 80 means aruco marker must be detected in 80% of the last pics to be displayed -->
 
<!--     <node pkg="aruco_detector_ocv" type="worldframe.py"
      name="worldframe_dynamic" output="screen"/>
    <node pkg="aruco_detector_ocv" type="listener_tf.py"
      name="listener_dynamic" output="screen"/> -->

   <!--   usb_cam node, as pixel_format use yuyv or mjpeg -->
<!--   <group if="$(arg use_camera)" >
  <node name="usb_cam" pkg="usb_cam" type="usb_cam_node" output="screen">
    <param name="video_device" value="/dev/video2" />
    <param name="image_width" value="640" />
    <param name="image_height" value="480" />
    <param name="pixel_format" value="yuyv" />
    <param name="camera_frame_id" value="usb_cam" />
    <param name="io_method" value="mmap"/>
  </node>
  </group> -->

    <!-- start bag file in a loop -->
<!--     <group if="$(arg use_bag)" >
        <node pkg="rosbag" type="play" name="rosbag_play" args="-l $(find aruco_detector_ocv)/data/$(arg bag_name).bag" />
    </group> -->


  <!-- D: [0.08269409104660361, -0.16570658726478218, 0.0013173694798400222, -0.0014110498528553663, 0.0], -->
  <!-- K: [519.734208757597, 0.0, 302.85096428096654, 0.0, 520.5923676830233, 252.63525359615707, 0.0, 0.0, 1.0], -->

<!-- 640x480 -->
<!--     <group if="$(eval arg('image_width') ==640)">
    <node name="camera_info" pkg="rostopic" type="rostopic"
      args="pub /camera/camera_info sensor_msgs/CameraInfo
     '{header: {seq: 0, stamp: {secs: 0, nsecs: 0}, frame_id: 'camera1'},
      height: 480, width: 640, distortion_model: 'plumb_bob',
      
      D: [1.0303738535136008e-01, -2.7013831868935040e-01, -2.5505644429536711e-03, -6.9928818760003142e-03, 1.6547444868453295e-01],
      K: [5.2036144200036904e+02, 0.0 , 3.0255876920712751e+02, 0.0 , 5.1834337983691967e+02, 2.4278187493920279e+02, 0.0, 0.0, 1.0],
      R: [1.0, 0.0, 0.0, 0.0, 1.0, 0.0, 0.0, 0.0, 1.0],
      P: [523.3435668945312, 0.0, 301.43917490757667, 0.0, 0.0, 526.1812744140625, 253.1422791119403, 0.0, 0.0, 0.0, 1.0, 0.0],
      binning_x: 0, binning_y: 0,
      roi: {x_offset: 0, y_offset: 0, height: 480, width: 640, do_rectify: false}}' -r 2"/>  
    </group>
 -->

<!-- 1920x1080  -->
<node if="$(eval arg('image_width') ==1920)" name="camera_info" pkg="rostopic" type="rostopic"
  args="pub /camera/camera_info sensor_msgs/CameraInfo
 '{header: {seq: 0, stamp: {secs: 0, nsecs: 0}, frame_id: 'camera1'},
  height: 1080, width: 1920, distortion_model: 'plumb_bob',
  D: [0.061957, -0.124832, 0.002573, -0.004753, 0.0],
  K: [1179.598752, 0.0, 928.099247, 0.0, 1177.000389, 558.635461, 0.0, 0.0, 1.0],
  R: [1.0, 0.0, 0.0, 0.0, 1.0, 0.0, 0.0, 0.0, 1.0],
  P: [1156.897217, 0.0, 908.598013, 0.0, 0.0, 1188.634277, 561.054263, 0.0, 0.0, 0.0, 1.0, 0.0],
  binning_x: 0, binning_y: 0,
  roi: {x_offset: 0, y_offset: 0, height: 1080, width: 1920, do_rectify: false}}' -r 2"/>
    <node pkg="aruco_detector_ocv" type="node" name="aruco_detect_node" output="screen">
        <param name="camera" type="string" value="$(arg camera)"/>
        <param name="camera_info" type="string" value="$(arg camera_info)"/>
        <param name="tf_prefix" type="string" value="$(arg tf_prefix)"/>
        <param name="show_detections" type="bool" value="$(arg show_detections)"/>
        <param name="marker_size" type="double" value="$(arg marker_size)"/>
        <param name="dictionary_name" type="string" value="$(arg dictionary_name)"/>
        <param name="image_width" type="double" value="$(arg image_width)"/>
        <param name="image_height" type="double" value="$(arg image_height)"/>
        <param name="blur_window_size" type="double" value="$(arg blur_window_size)"/>
        <param name="num_detected" type="double" value="$(arg num_detected)"/>
        <param name="min_prec_value" type="double" value="$(arg min_prec_value)"/>
    </node>

   <group if="$(arg use_rviz)" >
        <!-- <node pkg="rviz" type="rviz" name="markers_extrinsic" args="-d $(find aruco_detector_ocv)/config/aruco_opencv.rviz"/> -->
    </group>
</launch>
